### DISCUSSION TITLE
Changes to the editor's timing tab

### DISCUSSION THREAD

**Discussion by Fivoka** [ppy/osu#30295] [Ideas]
## Goal
The proposed changes aim to optimize the workflow of timing a song in Lazer and make it as smooth of a process as possible.

## Change list
_All of these features should only be available when using the timing tab.
Shortcuts are not final._

### Changes
- Remove the playback speed restrictions when using the timing tab, replace it with a warning notice instead and change the text color of “25%”, “50%”, “75%” to red.

### Basic Features
- `Ctrl+D`
Places a new timing point on the timeline cursor.

- `Shift+D`
Deletes the closest timing point to the left of the timeline cursor.

### New Features
- `Hold Ctrl`
Shows a button on every tick except for the one the timing point is on. When pressed, it adds a new timing point on that tick.

![timeline buttons](https://github.com/user-attachments/assets/807c4438-f324-4e8b-9e31-f88df4447812)

- `Hold Shift`
Shows a handle on every tick. When dragged, it adjusts the BPM of its corresponding timing point.
When dragging the tick that the timing point is on, its offset changes instead.

![timeline yellow cursor and handles](https://github.com/user-attachments/assets/dee6f777-5353-40aa-b8ce-0a11026748cd)

- `Hold Ctrl & Shift`
Shows some amalgamation of the two above buttons. It adjusts the BPM when dragged and adds a new timing point on that tick when released. When dragging the tick that the timing point is on, change its offset, as well as adjust the BPM of the previous timing point proportionally.

![timeline handle buttons](https://github.com/user-attachments/assets/63aabea1-d719-4166-99f1-a5d152399e5a)

_Only white ticks have larger buttons and handles._

- `Shift+Space`
Holding `Shift` alone turns the timeline cursor yellow. Pressing `Space` while `Shift` is held will keep the cursor yellow (`Shift` can be released) and play the song until the next white tick, at which point it will stop on its own and return to the normal cursor.
Note that with these shortcuts, holding `Shift` will also show handles on the ticks but won't interfere with that feature.

![timeline yellow cursor and handles](https://github.com/user-attachments/assets/dee6f777-5353-40aa-b8ce-0a11026748cd)

### Personal Issues
I don’t know if I should be waiting for a potentially coming redesign of the editor before starting to write any code, which I would love to do myself but I don’t really have any experience with C#. So if anyone is willing to spend some time teaching me, I’ll gladly listen, although it might take me a bit before I get the hang of it. But if you want to do it yourself, that's fine too.

In case anyone needs to contact me, my discord is “fivoka”, you'll have more of a chance of getting an immediate response there.

**Comment by OliBomby** [comment]
About the playback speed restrictions I'll add this https://github.com/ppy/osu/pull/28521#issuecomment-2208322454

I did some measurements myself back then to determine the offsets, the offsets are probably outdated since https://github.com/ppy/osu-framework/pull/6354

Peppy might be open to hardcoding some audio offsets for slower playback speeds, like how was done in stable, but it remains to be seen if these offsets are consistent between different platforms and hardware.

**Reply by Fivoka** [reply]
More of my opinion on the playback speed changes.

They still have other uses despite being unreliable to time with. Sometimes I just need to listen to the song slowed down because there's a lot going on in it, then once I figure out what's going on, I can go back to 100% playback rate and properly time the song.

The important thing then is to make sure that the user understands the implications of using non-100% playback rate, with a more detailed tooltip and a slightly scarier presentation.

![image](https://github.com/user-attachments/assets/de9b29df-554b-4be5-99a3-337c0d14987c)

**Reply by OliBomby** [reply]
I also prefer using lower playback speed when timing songs, so I want the restriction removed. What I'm saying is that it might be possible to get low playback speed with correct offset if we can determine the offsets and they happen to be constant. Having a correct offset low playback speed is much preferrable to having incorrect offset and a bunch of warnings. We should do some testing first before we determine which way to go.

**Comment by pha-Z** [comment]
its not just about audio-stretching-artifacts, but also about improved human "listening accuracy". its common sense that it is easier to correctly tell apart a sound as two distinct sounds (metronome vs music beat) the greater their time between them. and slowing down to 25% amplifies my human ability to do this by 4x. it dramatically improves timing accuracy.

also what i think should help to mitigate stretching-artifacts is the fact that to get to correct timing, i dont just listen to one single beat in the song one time. instead, by listening to the song at multiple different sections and multiple different playback speeds (i also listen to 100% playback speed as a sanity check because in the end it has to sound & play well at that speed anyways), i as a mapper effectively increase my data sample size and this should help reduce error variance.

please allow lower playback speed while timing. the tradeoff of improved human-listening-accuracy vs. audio-stretching-artifacts is worth it. stable osu had neither mappers nor players complain about how bad the accuracy of timed ranked maps are (because my guess is its actually pretty good - not bad)
